dependencies-in-practice.Rmd

```diff
@@ -0,0 +1,730 @@
+
+### What to export
+
+Export functions that you want other people to use.
+Exported functions must be documented, and you must be cautious when changing their interface --- other people are using them!
+Generally, it's better to export too little than too much.
+It's easy to start exporting something that you previously did not; it's hard to stop exporting a function because it might break existing code.
+Always err on the side of caution, and simplicity.
+It's easier to give people more functionality than it is to take away stuff they're used to.
+
+We believe that packages that have a wide audience should strive to do one thing and do it well.
+All functions in a package should be related to a single problem (or a set of closely related problems).
+Any functions not related to that purpose should not be exported.
+For example, most of our packages have a `utils.R` file (@sec-code-organising) that contains small utility functions that are useful internally, but aren't part of the core purpose of those packages.
+We don't export such functions.
+There are at least two reasons for this:
+
+-   Freedom to be less robust and less general.
+    A utility for internal use doesn't have to be implemented in the same way as a function used by others.
+    You just need to cover your own use case.
+
+-   Regrettable reverse dependencies.
+    You don't want people depending on your package for functionality and functions that are unrelated to its core purpose.
+
+That said, if you're creating a package for yourself, it's far less important to be this disciplined.
+Because you know what's in your package, it's fine to have a local "miscellany" package that contains a hodgepodge of functions that you find useful.
+But it is probably not a good idea to release such a package for wider use.
+
+### Re-exporting
+
+Sometimes you want to make something available to users of your package that is actually provided by one of your dependencies.
+When devtools was split into several smaller packages (@sec-setup-usage), many of the user-facing functions moved elsewhere.
+For usethis, the chosen solution was to list it in `Depends` (@sec-dependencies-imports-vs-depends), but that is not a good general solution.
+Instead, devtools now re-exports certain functions that actually live in a different package.
+
+Here is a blueprint for re-exporting an object from another package, using the `session_info()` function as our example:
+
+1.  List the package that hosts the re-exported object in `Imports` in `DESCRIPTION`.[^dependencies-in-practice-1]
+    In this case, the `session_info()` function is exported by the sessioninfo package.
+
+    ``` yaml
+    Imports:
+        sessioninfo
+    ```
+
+2.  In one of your `R/*.R` files, have a reference to the target function, preceded by roxygen tags for both importing and exporting.
+
+    ```{r}
+    #| eval: false
+    #' @export
+    #' @importFrom sessioninfo session_info
```

Comment:
Is it really true that you need `@importFrom`?

Reply:
Yes, you do. (And we do so in many, many places).

But I double-checked just to make sure:

```
** testing if installed package can be loaded from temporary location
Error: package or namespace load failed for ‘fiona’ in namespaceExport(ns, exports):
 undefined exports: session_info
Error: loading failed
Execution halted
ERROR: loading failed
```